docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+as described by the OCI container configuration file.
+4. **fixme** [Set up the container networking](https://github.com/clearcontainers/runtime/blob/master/documentation/architecture.md#networking).
+This must happen after all hooks are done as one of them is potentially setting
+the container networking namespace up.
+5. Create the virtual machine running the container process. The VM `systemd` instance will spawn the `agent` daemon.
+6. Wait for the `agent` to signal that it is ready.
+7. Send the pod creation command to the `agent`. The `agent` pod is the container process sandbox.
+8. Send the allocateIO command to the proxy, for getting the `agent` I/O sequence numbers described in step 2.
+9. Pass the `cc-proxy` socketpair file descriptor, and the I/O sequence numbers to the listening cc-shim process through the dedicated pipes.
+10. The `cc-shim` instance is put into a stopped state to prevent it from doing I/O before the container is started.
+
+![Docker create](create.png)
+
+At that point, the container sandbox is created in the virtual machine and `cc-shim` is stopped on the host.
+However the container process itself is not yet running as one needs to call `docker start` to actually start it.
+
+##### `start`
+
+On namespace containers `start` launches a traditional Linux container process in its own set of namespaces.
+With Clear Containers, the main task of `cc-runtime` is to create and start a container within the
+pod that got created during the `create` step. In practice, this means `cc-runtime` follows
+these steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know which pod
+we want to use to create and start the new container.
+2. `cc-runtime` sends an agent `NEWCONTAINER` command to create and start a new container in
+a given pod. The command is sent to `cc-proxy` who forwards it to the right agent instance running
+in the appropriate guest.
+3. `cc-runtime` resumes `cc-shim` so that it can now connect to the `cc-proxy` and acts as
+a signal and I/O streams proxy between `containerd-shim` and `cc-proxy`.
+
+##### `exec`
+
+`docker exec` allows one to run an additional command within an already running container.
+With Clear Containers, this translates into sending a `EXECCMD` command to the agent so
+that it runs a command into a running container belonging to a certain pod.
+All I/O streams from the executed command will be passed back to Docker through a newly
+created `cc-shim`.
+
+The `exec` code path is partly similar to the `create` one and `cc-runtime` goes through
+the following steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know which pod
+we want to use to run the `exec` command.
+2. `cc-runtime` sends the allocateIO command to the proxy, for getting the `agent` I/O sequence
+numbers for the `exec` command I/O streams.
```

Comment:
This has also changed with 3.0. We connect to the proxy and get a token back. We then create a shim with the given token and the shim connects to the proxy.